src/data/finset/powerset.lean

```diff
@@ -67,8 +67,63 @@ begin
     simp [finset.erase_eq_of_not_mem h, this] }
 end
 
+/-- For predicate `p` decidable on subsets, it is decidable whether `p` holds for any subset. -/
+def decidable_exists_of_decidable_subsets {s : finset α} {p : ∀ t ⊆ s, Prop}
+  (hu : ∀ t (h : t ⊆ s), decidable (p t h)) : decidable (∃ t (h : t ⊆ s), p t h) :=
+decidable_of_iff (∃ t (hs : t ∈ s.powerset), p t (mem_powerset.1 hs))
+⟨(λ ⟨t, _, hp⟩, ⟨t, _, hp⟩), (λ ⟨t, hs, hp⟩, ⟨t, mem_powerset.2 hs, hp⟩)⟩
+
+/-- For predicate `p` decidable on subsets, it is decidable whether `p` holds for every subset. -/
+def decidable_forall_of_decidable_subsets {s : finset α} {p : ∀ t ⊆ s, Prop}
+  (hu : ∀ t (h : t ⊆ s), decidable (p t h)) : decidable (∀ t (h : t ⊆ s), p t h) :=
+decidable_of_iff (∀ t (h : t ∈ s.powerset), p t (mem_powerset.1 h))
+⟨(λ h t hs, h t (mem_powerset.2 hs)), (λ h _ _, h _ _)⟩
+
+variables (p : finset α → Prop) [decidable_pred p]
+
+instance decidable_exists_of_subsets (s : finset α) : decidable (∃ t ⊆ s, p t) :=
+decidable_exists_of_decidable_subsets infer_instance
+
+instance decidable_forall_of_subsets (s : finset α) : decidable (∀ t ⊆ s, p t) :=
+decidable_forall_of_decidable_subsets infer_instance
+
 end powerset
 
+section ssubsets
+variables [decidable_eq α]
+
+/-- For `s` a finset, `s.ssubsets` is the finset comprising strict subsets of `s`. -/
+def ssubsets (s : finset α) : finset (finset α) :=
+erase (powerset s) s
+
+@[simp] lemma mem_ssubsets {s t : finset α} : t ∈ s.ssubsets ↔ t ⊂ s :=
+by rw [ssubsets, mem_erase, mem_powerset, ssubset_iff_subset_ne, and.comm]
+
+lemma empty_mem_ssubsets {s : finset α} (h : s.nonempty) : ∅ ∈ s.ssubsets :=
+by { rw [mem_ssubsets, ssubset_iff_subset_ne], exact ⟨empty_subset s, h.ne_empty.symm⟩, }
+
+/-- For predicate `p` decidable on ssubsets, it is decidable whether `p` holds for any ssubset. -/
+def decidable_exists_of_decidable_ssubsets {s : finset α} {p : ∀ t ⊂ s, Prop}
+  (hu : ∀ t (h : t ⊂ s), decidable (p t h)) : decidable (∃ t (h : t ⊂ s), p t h) :=
+decidable_of_iff (∃ t (hs : t ∈ s.ssubsets), p t (mem_ssubsets.1 hs))
+⟨(λ ⟨t, _, hp⟩, ⟨t, _, hp⟩), (λ ⟨t, hs, hp⟩, ⟨t, mem_ssubsets.2 hs, hp⟩)⟩
+
+/-- For predicate `p` decidable on ssubsets, it is decidable whether `p` holds for every ssubset. -/
+def decidable_forall_of_decidable_ssubsets {s : finset α} {p : ∀ t ⊂ s, Prop}
+  (hu : ∀ t (h : t ⊂ s), decidable (p t h)) : decidable (∀ t (h : t ⊂ s), p t h) :=
+decidable_of_iff (∀ t (h : t ∈ s.ssubsets), p t (mem_ssubsets.1 h))
+⟨(λ h t hs, h t (mem_ssubsets.2 hs)), (λ h _ _, h _ _)⟩
+
+variables (p : finset α → Prop) [decidable_pred p]
+
+instance decidable_exists_of_ssubsets (s : finset α) : decidable (∃ t ⊂ s, p t) :=
+decidable_exists_of_decidable_ssubsets infer_instance
+
+instance decidable_forall_of_ssubsets (s : finset α) : decidable (∀ t ⊂ s, p t) :=
+decidable_forall_of_decidable_ssubsets infer_instance
```

Comment:
```suggestion
/-- For predicate `p` decidable on ssubsets, it is decidable whether `p` holds for any ssubset. -/
instance decidable_exists_of_decidable_ssubsets {s : finset α} {p : ∀ t ⊂ s, Prop}
  [hu : ∀ t (h : t ⊂ s), decidable (p t h)] : decidable (∃ t (h : t ⊂ s), p t h) :=
decidable_of_iff (∃ t (hs : t ∈ s.ssubsets), p t (mem_ssubsets.1 hs))
⟨(λ ⟨t, _, hp⟩, ⟨t, _, hp⟩), (λ ⟨t, hs, hp⟩, ⟨t, mem_ssubsets.2 hs, hp⟩)⟩

/-- For predicate `p` decidable on ssubsets, it is decidable whether `p` holds for every ssubset. -/
instance decidable_forall_of_decidable_ssubsets {s : finset α} {p : ∀ t ⊂ s, Prop}
  [hu : ∀ t (h : t ⊂ s), decidable (p t h)] : decidable (∀ t (h : t ⊂ s), p t h) :=
decidable_of_iff (∀ t (h : t ∈ s.ssubsets), p t (mem_ssubsets.1 h))
⟨(λ h t hs, h t (mem_ssubsets.2 hs)), (λ h _ _, h _ _)⟩

-- TODO:  do you actually need these instances for your real use case?
variables (p : finset α → Prop) [decidable_pred p]

instance decidable_exists_of_ssubsets (s : finset α) : decidable (∃ t ⊂ s, p t) :=
infer_instance

instance decidable_forall_of_ssubsets (s : finset α) : decidable (∀ t ⊂ s, p t) :=
infer_instance
```

Reply:
More accurately, I am unable to channel a proof of `∀ t, t ⊂ s → decidable (p t)` through the typeclass system. Unless you can prove me wrong I will continue to believe that it has to be provided explicitly. Try this:
```
variables {s : finset ℕ} {p : finset ℕ → Prop}
instance p_dec_inst : ∀ t, t ⊂ s → decidable (p t) := sorry
def whatevs : decidable (∃ t, t ⊂ s → p t) := by apply_instance
```

Reply:
Does my proposed patch above make things worse?

Reply:
Sorry yes, I should have said that that code will fail in the context of your patch.

Reply:
And it succeeds in the context of yours?

Reply:
It doesn't succeed with my version. It doesn't succeed with your version. As far as I am aware, it doesn't succeed with any version. Because, as I keep having to say over and over, one cannot channel a proof like `p_dec_inst` through the typeclass system. It must be provided explicitly.

Reply:
I'm confused because the `by apply_instance` in my suggested patch seemed to work when I tried it.

Reply:
Hi!  I did say somewhere up the page that it was the dependence on `s` (more specifically the dependence on `t ⊂ s`) that does not flow through the typeclass system. So yes, the more straightforward `decidable_pred p` is unproblematic.

Reply:
I'll try to write more clearly tomorrow. I guess there are trade-offs here which depend on the precise use case. Very little in it either way! I just didn't see the point of requiring the use of @ notation any more than is strictly necessary.

Reply:
I think I understand your problem now, sorry. The problem is actually a little more nuanced:
```lean
section ok

variables {s : finset ℕ} {p : ∀ t, t ⊂ s → Prop}
instance p_dec_inst : ∀ t (H : t ⊂ s), decidable (p t H) := sorry
example : ∀ t (H : t ⊂ s), decidable (p t H) := by apply_instance

end ok

section fails

variables {s : finset ℕ} {p : ∀ t : finset ℕ, Prop}  -- only change: p does not depend on H
instance p_dec_inst_fail : ∀ t (H : t ⊂ s), decidable (p t) := sorry
example : ∀ t (H : t ⊂ s), decidable (p t) := by apply_instance

end fails
```
It appears typeclass search is unable to enter the `H` binder when it's not used within the term

Reply:
Which suggests to me that you should remove the `decidable_pred` instances (as lean can follow those from the instances I suggest), and add these lemmas isntead:
```lean
/-- A version of `finset.decidable_forall_of_decidable_ssubsets` with a non-dependent `p`. Typeclass inference cannot find `hu` here, so this is not an instance. -/
lemma decidable_forall_of_decidable_ssubsets' {s : finset α} {p : finset α → Prop}
  (hu : Π t (h : t ⊂ s), decidable (p t)) : decidable (∀ t (h : t ⊂ s), p t) :=
@finset.decidable_forall_of_decidable_ssubsets _ _ _ _ hu
```

Reply:
OK sure why not? It doesn't save lines or avoid `@` notation but it does cater better for the case of an instance of `∀ t (h : t ⊆ s), decidable (p t h)` (although I suspect you're most likely to find such a term in the middle of an induction so you'll have to use `@` or `haveI` anyway). I'll make the changes now.

Reply:
Is it ok to write lemma for a decidable? I've never understood what is the difference between lemma and def.

Reply:
No you're right, it should be a def